src/intrinsics/RegExpPrototype.mjs

```diff
@@ -118,37 +124,92 @@ function RegExpBuiltinExec(R, S) {
   const n = r.captures.length - 1;
   Assert(n < (2 ** 32) - 1);
   const A = X(ArrayCreate(new Value(n + 1)));
-  // Assert: The value of A's "length" property is n + 1.
+  Assert(X(Get(A, new Value('length'))).numberValue() === n + 1);
   X(CreateDataProperty(A, new Value('index'), lastIndex));
   X(CreateDataProperty(A, new Value('input'), S));
-  const matchedSubstr = S.stringValue().substring(lastIndex.numberValue(), e);
-  X(CreateDataProperty(A, new Value('0'), new Value(matchedSubstr)));
 
-  let groups;
-  if (R.parsedRegExp.groupSpecifiers.size > 0) {
-    groups = OrdinaryObjectCreate(Value.null);
+  const capturingParens = R.parsedRegExp.capturingParens;
+
+  if (surroundingAgent.feature('RegExpMatchIndices')) {
+    const indices = [];
+    const match = new MatchRecord(lastIndex.numberValue(), e);
+    indices.push(match);
+    const matchedValue = X(GetMatchString(S, match));
+    X(CreateDataProperty(A, new Value('0'), matchedValue));
+
+    let groups;
+    let groupNames;
+    if (R.parsedRegExp.groupSpecifiers.size > 0) {
+      groups = OrdinaryObjectCreate(Value.null);
+      groupNames = [];
+      // TODO: add this to spec text.
+      groupNames.push(Value.undefined);
```

Comment:
Without this, the access to the values in groupNames will be off by one in `MakeIndicesArray`